File: arrays/peeking-iterator/peeking-iterator.java
import java.util.Iterator;

// Java Iterator interface reference:
// https://docs.oracle.com/javase/8/docs/api/java/util/Iterator.html

class PeekingIterator implements Iterator<Integer> {
  Iterator<Integer> iterator;
  Integer head;

  public PeekingIterator(Iterator<Integer> iterator) {
    // initialize any member here.
    this.iterator = iterator;
    if (this.iterator.hasNext()) {
      head = this.iterator.next();
    }
  }

  // Returns the next element in the iteration without advancing the iterator.
  public Integer peek() {
    return head;
  }

  // hasNext() and next() should behave the same as in the Iterator interface.
  // Override them if needed.
  @Override
  public Integer next() {
    Integer res = head;
    if (iterator.hasNext()) {
      head = iterator.next();
    } else {
      head = null;
    }
    return res;
  }

  @Override
  public boolean hasNext() {
    return head != null;
  }
}

File: arrays/peeking-iterator/peeking-iterator.go
package peekingiterator

/*   Below is the interface for Iterator, which is already defined for you.
 *
 *   type Iterator struct {
 *
 *   }
 *
 *   func (this *Iterator) hasNext() bool {
 *		// Returns true if the iteration has more elements.
 *   }
 *
 *   func (this *Iterator) next() int {
 *		// Returns the next element in the iteration.
 *   }
 */

type PeekingIterator struct {
	iterator *Iterator
	head     *int
}

func Constructor(iter *Iterator) *PeekingIterator {

	var res PeekingIterator

	if iter.hasNext() {
		val := iter.next()
		res.head = &val
	}
	res.iterator = iter

	return &res
}

func (this *PeekingIterator) hasNext() bool {
	return this.head != nil
}

func (this *PeekingIterator) next() int {
	res := this.head
	if this.iterator.hasNext() {
		val := this.iterator.next()
		this.head = &val
	} else {
		this.head = nil
	}

	return *res
}

func (this *PeekingIterator) peek() int {
	return *this.head
}
